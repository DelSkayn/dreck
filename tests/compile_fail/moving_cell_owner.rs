use dreck::*;

struct DerefDrop<'gc, 'cell> {
    ptr: Gc<'gc, 'cell, i32>,
    owner: CellOwner<'cell>,
}

unsafe impl<'gc, 'cell> Trace for DerefDrop<'gc, 'cell> {
    fn needs_trace() -> bool
    where
        Self: Sized,
    {
        true
    }

    fn trace(&self, trace: Tracer) {
        trace.mark(self.ptr);
    }
}

unsafe impl<'a, 'gc, 'cell> Rebind<'a> for DerefDrop<'gc, 'cell> {
    type Output = DerefDrop<'gc, 'cell>;
}

impl<'gc, 'cell> Drop for DerefDrop<'gc, 'cell> {
    fn drop(&mut self) {
        // Accessing ptr in drop, very unsafe.
        self.ptr.borrow(&self.owner);
    }
}

fn main() {
    new_root!(owner, root);
    let ptr = root.add(0);
    let v = DerefDrop { ptr, owner: *owner };
    root.add(v);
    // ptr is collected here which will acces its contained pointer
    // which is UB.
}
