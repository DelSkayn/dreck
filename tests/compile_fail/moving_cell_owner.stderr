error[E0507]: cannot move out of `*owner` which is behind a mutable reference
  --> tests/compile_fail/moving_cell_owner.rs:35:37
   |
35 |     let v = DerefDrop { ptr, owner: *owner };
   |                                     ^^^^^^ move occurs because `*owner` has type `dreck::CellOwner<'_>`, which does not implement the `Copy` trait
